packages/cli/src/utils/deploy-v2.ts

```diff
@@ -0,0 +1,280 @@
+        baseRoute + tableConfig.route,
+        tableName,
+        Object.keys(tableConfig.schema)
+      );
+
+      console.log(chalk.green("Registered table", tableName, "at", baseRoute + tableConfig.route));
+    })
+  );
+
+  // Register systems (using forEach instead of for..of to avoid blocking on async calls)
+  promises.push(
+    ...Object.entries(mudConfig.systems).map(async ([systemName, systemConfig]) => {
+      console.log(chalk.blue("Registering system", systemName, "at", baseRoute + systemConfig.route));
+
+      // Register system route
+      const routeFragments = systemConfig.route.substring(1).split("/"); // Split route into fragments (skip leading slash)
+      const lastRouteFragment = toRoute(routeFragments.pop()); // Register last fragment as part of call to registerSystem
+      await registerNestedRoute(WorldContract, baseRoute, routeFragments);
+
+      // Register system at route
+      await safeTxExecute(
+        WorldContract,
+        "registerSystem",
+        baseRoute,
+        lastRouteFragment,
+        await contractPromises[systemName],
+        systemConfig.openAccess
+      );
+
+      console.log(chalk.green("Registered system", systemName, "at", baseRoute + systemConfig.route));
+    })
+  );
+
+  // Wait for routes to be registered before granting access to them
+  await Promise.all(promises);
+
+  // Grant access to systems
+  for (const [systemName, systemConfig] of Object.entries(mudConfig.systems)) {
+    const systemRoute = baseRoute + systemConfig.route;
+
+    // Grant access to addresses
+    promises.push(
+      ...systemConfig.accessListAddresses.map(async (address) => {
+        console.log(chalk.blue(`Grant ${address} access to ${systemName} (${systemRoute})`));
+        await safeTxExecute(WorldContract, "grantAccess", systemRoute, address);
+        console.log(chalk.green(`Granted ${address} access to ${systemName} (${systemRoute})`));
+      })
+    );
+
+    // Grant access to other systems
+    promises.push(
+      ...systemConfig.accessListSystems.map(async (granteeSystem) => {
+        console.log(chalk.blue(`Grant ${granteeSystem} access to ${systemName} (${systemRoute})`));
+        await safeTxExecute(WorldContract, "grantAccess", systemRoute, await contractPromises[granteeSystem]);
+        console.log(chalk.green(`Granted ${granteeSystem} access to ${systemName} (${systemRoute})`));
+      })
+    );
+  }
+
+  // Await all promises
+  await Promise.all(promises);
+
+  // Execute postDeploy forge script
+  const postDeployPath = path.join(await getScriptDirectory(), postDeployScript + ".s.sol");
+  if (existsSync(postDeployPath)) {
+    console.log(chalk.blue(`Executing post deploy script at ${postDeployPath}`));
+    console.log(
+      await forge(
+        "script",
+        postDeployScript,
+        "--sig",
+        "run(address)",
+        await contractPromises.World,
+        "--rpc-url",
+        rpc,
+        "--private-key",
+        privateKey,
```

Comment:
Unfortunately foundry doesn't seem to pick up the `PRIVATE_KEY` environment variable by default to replace the `--private-key` argument, but only provides a `vm` method to load it in foundry scripts. But since we use ethers to deploy / call contracts it doesn't really matter. I've replaced our `privateKey` cli param with loading the `PRIVATE_KEY` environment variable and never logging it in clear text.

Reply:
And also moved to using the rpc url from `foundry.toml` (and adding a `profile` param to our cli, which acts both as selector for the foundry profile, and we can use it to have default logic based on the profile, eg automatically funding the wallet when deploying to the lattice testnet).

Reply:
Looks like `FOUNDRY_PROFILE` is ignored in `cast block-number`, maybe also in other cast commands? Reported here: https://github.com/foundry-rs/foundry/issues/4474, will still load the rpc url from the foundry.toml but also pass as an explicit argument to cast for now.